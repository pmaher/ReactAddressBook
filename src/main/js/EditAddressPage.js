import React from 'react';
import axios from 'axios';
import { SelectState } from './SelectState';
import { withRouter } from 'react-router-dom';

export class EditAddressPage extends React.Component {
	render() {

		return (
			<div className="cover-container">
				<header className="masthead mb-auto text-center">
					<div className="inner">
						<h3>Edit Address</h3>
					</div>
				</header>
			
				<EditAddressWithRouter addressId={ this.props.match.params.addressId }/>

			</div>		
		)
	}
}

export class EditAddress extends React.Component {
	
	constructor(props) {
		super(props);
		
		this.state = {}
		this.handleAddressElementChange = this.handleAddressElementChange.bind(this);
		this.saveAddress = this.saveAddress.bind(this);
	}
	
	componentDidMount() {
		//this is the param passed from the url
		const addressId = this.props.addressId;
		axios.get('/api/address/' + addressId).then((response) => {
			this.setState({
				address: response.data
			})
		});
	}
	
	
    handleAddressElementChange(e) {
    	const newAddress = Object.assign(this.state.address, {[e.target.id]: e.target.value});
    	console.log(newAddress);
        this.setState({
        	address: newAddress
        });
    }
    
    saveAddress() {
		axios.put('/api/address/' + this.state.address.id, this.state.address).then((response) => {
			this.props.history.goBack();
		});
    }
	
	render() {
		const address = this.state.address;
		if(address) {
			return (
				<div className="container">
					<div className="col-12 col-md-9">
						<fieldset className="form-group">
							<div className="form-group">
						    	<label htmlFor="firstName">First Name:</label>
						    	<input id="firstName" type="text" className="form-control" placeholder="First Name" 
						    			value={ this.state.address.firstName } 
						    			onChange={ this.handleAddressElementChange } />
						    </div>
							<div className="form-group">
						    	<label htmlFor="firstName">Last Name:</label>
						    	<input id="lastName" type="text" className="form-control" placeholder="Last Name" 
						    		value={ address.lastName } 
						    		onChange={ this.handleAddressElementChange }/>
						    </div>
							<div className="form-group">
						    	<label htmlFor="address">Address:</label>
						    	<input id="address" type="text" className="form-control" placeholder="Address" 
						    		value={ address.address }
						    		onChange={ this.handleAddressElementChange } />
						    </div>
							<div className="form-group">
						    	<label htmlFor="address2">Address 2:</label>
						    	<input id="address2" type="text" className="form-control" placeholder="Address 2" 
						    		value={ address.address2 } 
						    		onChange={ this.handleAddressElementChange } />
						    </div>
							<div className="form-group">
						    	<label htmlFor="city">City:</label>
						    	<input id="city" type="text" className="form-control" placeholder="City" 
						    		value={ address.city }
						    		onChange={ this.handleAddressElementChange } />
						    </div>
							<div className="form-group">
						    	<label htmlFor="state">State:</label>
						    	<SelectState id="state" className="form-control" placeholder="State" 
						    		selectedState={ address.state }
						    		onChange={ this.handleAddressElementChange } />
						    </div>
							<div className="form-group">
						    	<label htmlFor="zip">Zip:</label>
						    	<input id="zipcode" type="text" className="form-control" placeholder="Zip" 
						    		value={ address.zipcode } 
						    		onChange={ this.handleAddressElementChange } />
						    </div>
							<div className="form-group">
						    	<label htmlFor="email">Email:</label>
						    	<input id="email" type="text" className="form-control" placeholder="Email" 
						    		value={ address.email }
						    		onChange={ this.handleAddressElementChange } />
						    </div>
							<div className="form-group">
						    	<label htmlFor="phone">Phone:</label>
						    	<input id="phone" type="text" className="form-control" placeholder="Phone" 
						    		value={ address.phone } 
						    		onChange={ this.handleAddressElementChange } />
						    </div>
						</fieldset>
						<div className="btn-toolbar" role="toolbar" aria-label="Toolbar with button groups">
							<CancelButton>Cancel</CancelButton>
							<SaveAddressButton saveAddress={ this.saveAddress } />
						</div>
					</div>
				</div>
				)
		} else {
			return (
				<div className="container">	
					No Addresses Found.  Please go back to select an existing address or create a new one.
					<CancelButton >Back</CancelButton>
				</div>
			)
		}
	}
}

const EditAddressWithRouter = withRouter(EditAddress);

const CancelButton = withRouter( (props) => (
		<button className="btn btn-info btn-space" onClick={ ()=> props.history.goBack() }>{ props.children }</button>
	)
);

const SaveAddressButton = (props) => (
	<button className="btn btn-primary btn-space" onClick={ props.saveAddress }>Save</button>
);