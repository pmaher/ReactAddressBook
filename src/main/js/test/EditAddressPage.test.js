require('./helpers/enzyme-setup');
import React from 'react';
import { EditAddress } from '../EditAddressPage';
import axios from 'axios';
import renderer from 'react-test-renderer';
//the MemoryRouter is used to get around issues around React context
import { MemoryRouter } from 'react-router-dom';
jest.mock('axios');

describe('EditAddress', () => {
	it('Should make ajax request to pull the appropriate address information', done => {
		const mockAddressData = { 
									data: { 
										"address": "2693 Commerce St.",
										"address2":"",
										"city":"Northridge",
										"email":"Linda.Douglas@gmail.com",
										"firstName":"Linda",
										"lastName":"Douglas",
										"id":3,
										"phone":"(805) 333-9999",
										"state":"CA",
										"zipcode":"91368"
									}
								};
		axios.get.mockResolvedValue(mockAddressData);
    	const wrapper = mount(<MemoryRouter><EditAddress addressId="3" /></MemoryRouter>);

    	//queue this assertion callback behind the axios mock promise being resolved
    	setImmediate(() => {
        	//this update is needed since the enzyme wrapper is immutable as of v.3
        	wrapper.update();
        	//using wrapper.debug() here to avoid clutter from the MemoryRouter rendering excessive attributes
        	expect(wrapper.debug()).toMatchSnapshot();
        	done();
    	});
	});
})