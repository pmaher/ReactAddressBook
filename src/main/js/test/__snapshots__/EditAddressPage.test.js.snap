// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditAddress Should make ajax request to pull the appropriate address information 1`] = `
"<MemoryRouter>
  <Router history={{...}}>
    <EditAddress addressId=\\"3\\">
      <div className=\\"container\\">
        <div className=\\"col-12 col-md-9\\">
          <fieldset className=\\"form-group\\">
            <div className=\\"form-group\\">
              <label htmlFor=\\"firstName\\">
                First Name:
              </label>
              <input id=\\"firstName\\" type=\\"text\\" className=\\"form-control\\" placeholder=\\"First Name\\" value=\\"Linda\\" onChange={[Function: bound handleAddressElementChange]} />
            </div>
            <div className=\\"form-group\\">
              <label htmlFor=\\"firstName\\">
                Last Name:
              </label>
              <input id=\\"lastName\\" type=\\"text\\" className=\\"form-control\\" placeholder=\\"Last Name\\" value=\\"Douglas\\" onChange={[Function: bound handleAddressElementChange]} />
            </div>
            <div className=\\"form-group\\">
              <label htmlFor=\\"address\\">
                Address:
              </label>
              <input id=\\"address\\" type=\\"text\\" className=\\"form-control\\" placeholder=\\"Address\\" value=\\"2693 Commerce St.\\" onChange={[Function: bound handleAddressElementChange]} />
            </div>
            <div className=\\"form-group\\">
              <label htmlFor=\\"address2\\">
                Address 2:
              </label>
              <input id=\\"address2\\" type=\\"text\\" className=\\"form-control\\" placeholder=\\"Address 2\\" value=\\"\\" onChange={[Function: bound handleAddressElementChange]} />
            </div>
            <div className=\\"form-group\\">
              <label htmlFor=\\"city\\">
                City:
              </label>
              <input id=\\"city\\" type=\\"text\\" className=\\"form-control\\" placeholder=\\"City\\" value=\\"Northridge\\" onChange={[Function: bound handleAddressElementChange]} />
            </div>
            <div className=\\"form-group\\">
              <label htmlFor=\\"state\\">
                State:
              </label>
              <SelectState id=\\"state\\" className=\\"form-control\\" placeholder=\\"State\\" selectedState=\\"CA\\" onChange={[Function: bound handleAddressElementChange]}>
                <select id=\\"state\\" className=\\"form-control\\" onChange={[Function: bound _handleChange]} value=\\"CA\\">
                  <option value=\\"AL\\">
                    Alabama
                  </option>
                  <option value=\\"AK\\">
                    Alaska
                  </option>
                  <option value=\\"AS\\">
                    American Samoa
                  </option>
                  <option value=\\"AZ\\">
                    Arizona
                  </option>
                  <option value=\\"AR\\">
                    Arkansas
                  </option>
                  <option value=\\"CA\\">
                    California
                  </option>
                  <option value=\\"CO\\">
                    Colorado
                  </option>
                  <option value=\\"CT\\">
                    Connecticut
                  </option>
                  <option value=\\"DE\\">
                    Delaware
                  </option>
                  <option value=\\"DC\\">
                    District Of Columbia
                  </option>
                  <option value=\\"FM\\">
                    Federated States Of Micronesia
                  </option>
                  <option value=\\"FL\\">
                    Florida
                  </option>
                  <option value=\\"GA\\">
                    Georgia
                  </option>
                  <option value=\\"GU\\">
                    Guam
                  </option>
                  <option value=\\"HI\\">
                    Hawaii
                  </option>
                  <option value=\\"ID\\">
                    Idaho
                  </option>
                  <option value=\\"IL\\">
                    Illinois
                  </option>
                  <option value=\\"IN\\">
                    Indiana
                  </option>
                  <option value=\\"IA\\">
                    Iowa
                  </option>
                  <option value=\\"KS\\">
                    Kansas
                  </option>
                  <option value=\\"KY\\">
                    Kentucky
                  </option>
                  <option value=\\"LA\\">
                    Louisiana
                  </option>
                  <option value=\\"ME\\">
                    Maine
                  </option>
                  <option value=\\"MH\\">
                    Marshall Islands
                  </option>
                  <option value=\\"MD\\">
                    Maryland
                  </option>
                  <option value=\\"MA\\">
                    Massachusetts
                  </option>
                  <option value=\\"MI\\">
                    Michigan
                  </option>
                  <option value=\\"MN\\">
                    Minnesota
                  </option>
                  <option value=\\"MS\\">
                    Mississippi
                  </option>
                  <option value=\\"MO\\">
                    Missouri
                  </option>
                  <option value=\\"MT\\">
                    Montana
                  </option>
                  <option value=\\"NE\\">
                    Nebraska
                  </option>
                  <option value=\\"NV\\">
                    Nevada
                  </option>
                  <option value=\\"NH\\">
                    New Hampshire
                  </option>
                  <option value=\\"NJ\\">
                    New Jersey
                  </option>
                  <option value=\\"NM\\">
                    New Mexico
                  </option>
                  <option value=\\"NY\\">
                    New York
                  </option>
                  <option value=\\"NC\\">
                    North Carolina
                  </option>
                  <option value=\\"ND\\">
                    North Dakota
                  </option>
                  <option value=\\"MP\\">
                    Northern Mariana Islands
                  </option>
                  <option value=\\"OH\\">
                    Ohio
                  </option>
                  <option value=\\"OK\\">
                    Oklahoma
                  </option>
                  <option value=\\"OR\\">
                    Oregon
                  </option>
                  <option value=\\"PW\\">
                    Palau
                  </option>
                  <option value=\\"PA\\">
                    Pennsylvania
                  </option>
                  <option value=\\"PR\\">
                    Puerto Rico
                  </option>
                  <option value=\\"RI\\">
                    Rhode Island
                  </option>
                  <option value=\\"SC\\">
                    South Carolina
                  </option>
                  <option value=\\"SD\\">
                    South Dakota
                  </option>
                  <option value=\\"TN\\">
                    Tennessee
                  </option>
                  <option value=\\"TX\\">
                    Texas
                  </option>
                  <option value=\\"UT\\">
                    Utah
                  </option>
                  <option value=\\"VT\\">
                    Vermont
                  </option>
                  <option value=\\"VI\\">
                    Virgin Islands
                  </option>
                  <option value=\\"VA\\">
                    Virginia
                  </option>
                  <option value=\\"WA\\">
                    Washington
                  </option>
                  <option value=\\"WV\\">
                    West Virginia
                  </option>
                  <option value=\\"WI\\">
                    Wisconsin
                  </option>
                  <option value=\\"WY\\">
                    Wyoming
                  </option>
                </select>
              </SelectState>
            </div>
            <div className=\\"form-group\\">
              <label htmlFor=\\"zip\\">
                Zip:
              </label>
              <input id=\\"zipcode\\" type=\\"text\\" className=\\"form-control\\" placeholder=\\"Zip\\" value=\\"91368\\" onChange={[Function: bound handleAddressElementChange]} />
            </div>
            <div className=\\"form-group\\">
              <label htmlFor=\\"email\\">
                Email:
              </label>
              <input id=\\"email\\" type=\\"text\\" className=\\"form-control\\" placeholder=\\"Email\\" value=\\"Linda.Douglas@gmail.com\\" onChange={[Function: bound handleAddressElementChange]} />
            </div>
            <div className=\\"form-group\\">
              <label htmlFor=\\"phone\\">
                Phone:
              </label>
              <input id=\\"phone\\" type=\\"text\\" className=\\"form-control\\" placeholder=\\"Phone\\" value=\\"(805) 333-9999\\" onChange={[Function: bound handleAddressElementChange]} />
            </div>
          </fieldset>
          <div className=\\"btn-toolbar\\" role=\\"toolbar\\" aria-label=\\"Toolbar with button groups\\">
            <withRouter()>
              <Route>
                <Component match={{...}} location={{...}} history={{...}} staticContext={[undefined]}>
                  <button className=\\"btn btn-info btn-space\\" onClick={[Function: onClick]}>
                    Cancel
                  </button>
                </Component>
              </Route>
            </withRouter()>
            <SaveAddressButton saveAddress={[Function: bound saveAddress]}>
              <button className=\\"btn btn-primary btn-space\\" onClick={[Function: bound saveAddress]}>
                Save
              </button>
            </SaveAddressButton>
          </div>
        </div>
      </div>
    </EditAddress>
  </Router>
</MemoryRouter>"
`;
